### PR TITLE
arm: add support for softfp in arm vfp assembly files

### PR DESCRIPTION
**Pull request by ashwinyes**

### CODE REVIEW

**Comment by ashwinyes** [comment]
This PR adds support for softfp in all arm vfp assembly kernel files.

I have done only basic testing of this PR since I don't have an ARM machine at my disposal. All my testing were done on QEMU emulator.

Request owners of following issues to help test this PR and see whether the issue is resolved or not.

#1088 #1145 #1168 #1206 #1208 #1215 #1145 @misbullah @gangm @odieXin @ctgushiwei  @mariasmo @akira-miasato @gangm

**Comment by ashwinyes** [comment]
@tomaklutfu  Thanks for reviewing.

Yes. Those were wrong. I agree. Had corrected those in my local workspace. But still the GEMM answer was not correct. I didnt explore it much as these files were no longer required after all the changes were made in the assembly files.

Please see that I have deleted those files in the last commit.

**Comment by xianyi** [comment]
@ashwinyes Thank you for the PR.

Do I need to merge #1216 at first?

**Comment by ashwinyes** [comment]
No #1216 is not required. It can closed.

I would prefer these changes to get tested by someone working on android and get a "Tested" stamp on it. I have only tested it on QEMU emulator.

**Comment by xianyi** [comment]
We will test it this week

**Comment by mariasmo** [comment]
ashwinyes & xianyi,

i am trying to understand this request.

so i have been compiling openblas using the following command:

make TARGET=ARMV7 HOSTCC=gcc CC=arm-linux-androideabi-gcc NOFORTRAN=1 NUM_THREADS=32 libs

I believe it defaults to "hardfloat" right?

my understanding is that the PR removes the "hardfloat" by default? and can i use the same above mentioned command or do i need to use the following:
ARM_SOFTFP_ABI=1

and to test, do can i get the latest from develop branch? or should i wait until the PR is merged?

thanks

**Comment by ashwinyes** [comment]
Let me explain.

softfp and hard float ABI differs in the way the arguments are passed on to functions and how values are returned. If your project has full C files only, then compiler will take care of everything.

But issue is when you have hand written assembly code. The code in OpenBLAS was always assuming hard float ABI was used. So the code expects values to be certain registers according hard fp ABI which will not be the case when you use softfp ABI.

Now coming to your question. arm-linux-androideabi-gcc (also arm-linux-gnueabi-gcc) by default is **softfp ABI**. When you use this compiler to compile OpenBLAS code, the issue I mentioned above arises.

The PR doesn't remove anything. It just checks whether you are using a softfp or hard fp compliant and adjusts the arguments/return values accordingly.

So, basically you don't need to give **any** additional make parameters to OpenBLAS with the PR.

To test, you can use the branch https://github.com/ashwinyes/OpenBLAS/tree/develop_arm_softfp until this PR gets merged.

Hope I cleared your doubt.

**Comment by mariasmo** [comment]
ashwinyes

thanks for the explanation.

so prior to your changes OpenBLAS was assuming hard fp even though the default for arm-linux-androideabi-gcc  is softfp. Am i correct in my understanding?

and now with your changes , it will compile for only soffp and any other libraries i compile down the stream i shuould assume softfp and not hardfp, correct?

thanks

**Comment by ashwinyes** [comment]
Yes. You are correct.

**Comment by mariasmo** [comment]
thanks

**Comment by akira-miasato** [comment]
I'll be checking it today regarding #1168 and #1215.

**Comment by akira-miasato** [comment]
Tested and issue #1168 seems solved with Android NDK r15b, with speedups comparable to what I achieved with the dirty adaptation from #1215/#1216.

However, some changes regarding GCC/Clang support will be needed (NDK dropped GCC support). I should open a new issue for those things, since this PR focuses on the softfp/hard ABI issues.

**Comment by ashwinyes** [comment]
"fnmacs s3 , s1, s4" is a valid ARM instruction. If clang doesn't recognize, then its more of a clang issue than OpenBLAS issue. But I will try to see if I can workaround it.

**Comment by ashwinyes** [comment]
Looks like clang is not recognizing pre-UAL mnemonics. We will have to replace with UAL mnemonics.

http://infocenter.arm.com/help/index.jsp?topic=/com.arm.doc.dui0473c/CIHFAHFG.html

**Comment by ashwinyes** [comment]
@akira-miasato 
Addressed the clang issue in https://github.com/xianyi/OpenBLAS/pull/1226

**Comment by jxllh123456** [comment]
thanks a lot, India guy

**Comment by jxllh123456** [comment]
thanks a lot, sure enough, India's software industry has good superiority.
